htscodecs/tokenise_name3.c

```diff
@@ -1241,58 +1241,148 @@ static int64_t rans_decode(uint8_t *in, uint64_t in_len, uint8_t *out, uint64_t
+            /* ALPHA    */ {2, 129,1},
+            /* CHAR     */ {1, 0},
+            /* DIGITS0  */ {2, 128+8,0}, // size%4==0
+            /* DZLEN    */ {1, 0},
+            /* DUP      */ {1, 192+8},   // size%4==0
+            /* DIFF     */ {1, 128+8},   // size%4==0
+            /* DIGITS   */ {1, 192+8},   // size%4==0
+            /* DDELTA   */ {1, 0},
+            /* DDELTA0  */ {1, 128},
+            /* MATCH    */ {1, 0},
+            /* NOP      */ {1, 0},
+            /* END      */ {1, 0}
+        },
+
+        {   // -5
+            /* TYPE     */ {2, 192,0},
+            /* ALPHA    */ {4, 1,128,0,129},
+            /* CHAR     */ {1, 0},
+            /* DIGITS0  */ {2, 200,0},
+            /* DZLEN    */ {1, 0},
+            /* DUP      */ {1, 200},
+            /* DIFF     */ {2, 192,200},
+            /* DIGITS   */ {2, 132,201},
+            /* DDELTA   */ {1, 0},
+            /* DDELTA0  */ {1, 128},
+            /* MATCH    */ {1, 0},
+            /* NOP      */ {1, 0},
+            /* END      */ {1, 0}
+        },
+
+        {   // -7
+            /* TYPE     */ {3, 193,0,1},
+            /* ALPHA    */ {5, 128, 1,128,0,129},
+            /* CHAR     */ {2, 1,0},
+            /* DIGITS0  */ {2, 200,0},    // or 201,0
+            /* DZLEN    */ {1, 0},
+            /* DUP      */ {1, 201},
+            /* DIFF     */ {2, 192,200},  // or 192,201
+            /* DIGITS   */ {2, 132, 201}, // +bz2 here and -9
+            /* DDELTA   */ {1, 0},
+            /* DDELTA0  */ {1, 128},
+            /* MATCH    */ {1, 0},
+            /* NOP      */ {1, 0},
+            /* END      */ {1, 0}
+        },
+
+        {   // -9
+            /* TYPE     */ {6, 192,0,1,  65,  193,132},
+            /* ALPHA    */ {4, 132, 1, 0,129},
+            /* CHAR     */ {3, 1,0,192},
+            /* DIGITS0  */ {4, 201,0, 192,64},
+            /* DZLEN    */ {3, 0,128,1},
+            /* DUP      */ {1, 201},
+            /* DIFF     */ {3, 192,  201,65},
+            /* DIGITS   */ {6, 132, 201,1,  192,129,  193},
+            /* DDELTA   */ {3, 1,0,  192},
+            /* DDELTA0  */ {3, 192,1,  0},
+            /* MATCH    */ {1, 0},
+            /* NOP      */ {1, 0},
+            /* END      */ {1, 0}
+        },
+    };
+    // Minor tweak to level 3 DIGITS if arithmetic, to use O(201) instead.
+    if (use_arith) R[1][N_DIGITS][1]=201;
+
+    int *meth = R[level][type];
+
+    int last = 0, m;
+    uint8_t best_static[8192];
+    uint8_t *best_dat = best_static;
+    for (m = 1; m <= meth[0]; m++) {
         *out_len = olen;
 
-        if (in_len % 4 != 0 && (rmethods[level][m] & 8))
+        if (!use_arith && (meth[m] & 4))
+            meth[m] &= ~4;
+
+        if (in_len % 4 != 0 && (meth[m] & 8))
             continue;
 
+        last = 0;
         if (use_arith) {
-            if (arith_encode(in, in_len, out, out_len, rmethods[level][m]) < 0)
+            if (arith_encode(in, in_len, out, out_len, meth[m]) <0)
```

Comment:
It's now possible to leak memory here, if this (or `rans_encode()`) fails, and you allocated `best_dat` on the heap in an earlier iteration.

Reply:
Corrected.